From: MeiChia Chiu <meichia.chiu@mediatek.com>
Date: Tue, 6 Jun 2023 16:57:10 +0800
Subject: [PATCH] wifi: mt76: mt7996: fix in-band discovery runtime updates

Let the in-band discovery (FILS / unsolicited broadcast probe
response) templates be updated at runtime: drop the
fils_discovery.max_interval precondition and derive the enable flag
from the requested interval, so hostapd can reconfigure discovery
without a full interface restart.

Usage:
Enable FILS: hostapd_cli -i [interface] inband_discovery 2 20
Enable UBPR: hostapd_cli -i [interface] inband_discovery 1 20
Disable inband discovery: hostapd_cli -i [interface] inband_discovery 0 0

Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 mt7996/mcu.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3745,8 +3745,7 @@ int mt7996_mcu_beacon_inband_discov(stru
 	if (IS_ERR(rskb))
 		return PTR_ERR(rskb);
 
-	if (changed & BSS_CHANGED_FILS_DISCOVERY &&
-	    link_conf->fils_discovery.max_interval) {
+	if (changed & BSS_CHANGED_FILS_DISCOVERY) {
 		interval = link_conf->fils_discovery.max_interval;
 		skb = ieee80211_get_fils_discovery_tmpl(hw, vif);
 	} else if (changed & BSS_CHANGED_UNSOL_BCAST_PROBE_RESP &&
@@ -3781,7 +3780,7 @@ int mt7996_mcu_beacon_inband_discov(stru
 	discov->tx_type = !!(changed & BSS_CHANGED_FILS_DISCOVERY);
 	discov->tx_interval = interval;
 	discov->prob_rsp_len = cpu_to_le16(MT_TXD_SIZE + skb->len);
-	discov->enable = true;
+	discov->enable = !!(interval);
 	discov->wcid = cpu_to_le16(MT7996_WTBL_RESERVED);
 
 	buf = (u8 *)tlv + sizeof(*discov);
